internal/wal/wal.go

```diff
@@ -185,5 +269,30 @@ func (w *Wal) nextSegment() error {
 	// close and fsync the current active segment
 	// create new segment file
 	// fsync the directory
+	nextId := w.active.id + 1
+	if err := w.active.Sync(); err != nil {
+		return err
+	}
+	if err := w.active.Close(); err != nil {
+		return err
+	}
+	lsn := NewLSN(nextId, int64(w.opts.MaxSegmentSize), 0)
+	seg, err := createSegment(lsn, w.opts)
+	if err != nil {
+		return err
+	}
+	w.active = seg
 	return nil
 }
+
+func (w *Wal) writeData(data []byte) (int, error) {
+	w.lock.Lock()
```

Comment:
Move locks up to `wal.Write()` because you also have to ensure that w.active is updated atomically.